File: examples/node/offline-dpt.ts
// KNXyz example: offline DPT encode/decode (default-safe, print-only).
//
// This example does not connect to a KNX bus. It encodes and decodes KNX
// datapoint values from local bytes only, so it needs no network and no KNX
// hardware. The Node binding is a native addon: build it once before running
// this example (see examples/README.md).

import { encodeDpt, decodeDpt } from "@knxyz/knx";

function hex(payload: Uint8Array): string {
  return Array.from(payload, (b) => b.toString(16).padStart(2, "0")).join(" ");
}

// Scalar DPTs decode to a bare value; some (e.g. 17.001 in Node) decode to a
// tagged object { type, value }. Pull the scalar out for the round-trip check.
function scalarOf(decoded: unknown): unknown {
  if (decoded !== null && typeof decoded === "object" && "value" in (decoded as object)) {
    return (decoded as { value: unknown }).value;
  }
  return decoded;
}

const cases: Array<[string, string, unknown]> = [
  ["1.001", "switch (boolean)", true],
  ["9.001", "temperature (degC, Float16)", 21.0],
  ["9.001", "temperature negative (degC)", -5.5],
  ["5.010", "counter (raw 0-255)", 128],
  ["17.001", "scene number", 7],
];

let ok = true;
for (const [dpt, label, value] of cases) {
  const payload = encodeDpt(dpt, value);
  const decoded = decodeDpt(dpt, payload);
  const matched = scalarOf(decoded) === value;
  ok = ok && matched;
  console.log(
    `DPT ${dpt.padEnd(7)} ${label.padEnd(28)} -> bytes=[${hex(payload)}] -> ` +
      `decoded=${JSON.stringify(decoded)} ${matched ? "OK" : "MISMATCH"}`,
  );
}

console.log(`offline round-trip: ${ok ? "all OK" : "FAILED"}`);
if (!ok) {
  process.exit(1);
}

File: examples/python/offline_dpt.py
"""Offline DPT encode/decode example (default-safe, print-only).

This example does not connect to a KNX bus. It encodes and decodes KNX datapoint
values from local bytes only, so it needs no network and no KNX hardware. Run it
from a source checkout with the locally-built package:
  python examples/python/offline_dpt.py
"""

import sys

from knxyz import dpt


def _hex(payload: bytes) -> str:
    return " ".join("%02x" % b for b in payload)


def main() -> int:
    # Scalar DPTs round-trip to a bare Python value (bool / int / float).
    cases = [
        ("1.001", "switch (boolean)", True),
        ("9.001", "temperature (degC, Float16)", 21.0),
        ("9.001", "temperature negative (degC)", -5.5),
        ("5.010", "counter (raw 0-255)", 128),
        ("17.001", "scene number", 7),
    ]
    ok = True
    for dpt_id, label, value in cases:
        payload = dpt.encode(dpt_id, value)
        decoded = dpt.decode(dpt_id, payload)
        matched = decoded == value
        ok = ok and matched
        print(
            "DPT %-7s %-28s value=%-6r -> bytes=[%s] -> decoded=%r %s"
            % (dpt_id, label, value, _hex(payload), decoded, "OK" if matched else "MISMATCH")
        )
    print("offline round-trip:", "all OK" if ok else "FAILED")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())
